<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Precourse – Population Genomics In Practice H25</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./assets/logos/nbislogo-green.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Population Genomics In Practice H25</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./syllabus.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./info.html"> 
<span class="menu-text">Practical info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./precourse.html" aria-current="page"> 
<span class="menu-text">Precourse</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#dardel-compute-resources" id="toc-dardel-compute-resources" class="nav-link active" data-scroll-target="#dardel-compute-resources">Dardel compute resources</a>
  <ul class="collapse">
<li><a href="#naiss-account" id="toc-naiss-account" class="nav-link" data-scroll-target="#naiss-account">NAISS account</a></li>
  <li><a href="#dardel-tutorials" id="toc-dardel-tutorials" class="nav-link" data-scroll-target="#dardel-tutorials">Dardel tutorials</a></li>
  </ul>
</li>
  <li><a href="#sec-software-installation" id="toc-sec-software-installation" class="nav-link" data-scroll-target="#sec-software-installation">Software installation</a></li>
  <li><a href="#suggested-readings" id="toc-suggested-readings" class="nav-link" data-scroll-target="#suggested-readings">Suggested readings</a></li>
  <li>
<a href="#resources" id="toc-resources" class="nav-link" data-scroll-target="#resources">Resources</a>
  <ul class="collapse">
<li><a href="#literature" id="toc-literature" class="nav-link" data-scroll-target="#literature">Literature</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Precourse</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="dardel-compute-resources" class="level2"><h2 class="anchored" data-anchor-id="dardel-compute-resources">Dardel compute resources</h2>
<!-- markdownlint-disable MD013 -->
<p>Computer exercises can be run on the <a href="https://www.pdc.kth.se/hpc-services/computing-systems/dardel-hpc-system/dardel-1.1043529">Dardel HPC cluster</a> at Stockholm maintained by <a href="https://www.pdc.kth.se/">PDC</a>. However, you can also opt to run exercises on your own computer (see <a href="#sec-software-installation">software installation</a> below).</p>
<!-- markdownlint-enable MD013 -->
<section id="naiss-account" class="level3"><h3 class="anchored" data-anchor-id="naiss-account">NAISS account</h3>
<!-- markdownlint-disable MD013 -->
<p>You will need a <a href="https://support.pdc.kth.se/doc/getting_access/get_access/#applying-for-an-account">PDC account</a> to run exercises on Dardel. This is easiest done by <a href="https://supr.naiss.se/person/register/">applying for a SUPR account</a> and request access to the PGIP compute project <code>pg_pgip_2025</code>.</p>
<!-- markdownlint-enable MD013 -->
</section><section id="dardel-tutorials" class="level3"><h3 class="anchored" data-anchor-id="dardel-tutorials">Dardel tutorials</h3>
<p>Look at <a href="https://support.pdc.kth.se/doc/site_map/" class="uri">https://support.pdc.kth.se/doc/site_map/</a>, in particular <a href="https://support.pdc.kth.se/doc/basics/quickstart/#how-to-log-in" class="uri">https://support.pdc.kth.se/doc/basics/quickstart/#how-to-log-in</a> for information on how to connect to and work on Dardel.</p>
</section></section><section id="sec-software-installation" class="level2"><h2 class="anchored" data-anchor-id="sec-software-installation">Software installation</h2>
<p>Although we will mainly work on Dardel, some exercises, notably the ones that utilise <a href="https://jupyter.org/">Jupyter</a> notebooks, benefit from working in a local compute environment on your computer. We will use the <a href="https://pixi.sh/latest/">pixi</a> package manager to install necessary requirements from the package repositories <a href="https://bioconda.github.io/">bioconda</a> and <a href="https://conda-forge.org/">conda-forge</a>.</p>
<p>To start using pixi, follow the <a href="https://pixi.sh/latest/installation/">install instructions</a> to install. For a given exercise, software dependencise will be provided in a <a href="https://pixi.sh/latest/reference/pixi_manifest/#manifest-discovery">pixi manifest</a>, an example of which is shown here:</p>
<pre lang="text"><code>[workspace]
channels = ["conda-forge", "bioconda"]
name = "variant-calling"

[dependencies]
fastqc = "&gt;=0.12.1,&lt;0.13"</code></pre>
<p>Copy the contents to a file <code>pixi.toml</code> in an appropriately named directory (e.g., <code>variant-calling</code>), <code>cd</code> to directory and activate environment with <code>pixi shell</code>. As mentioned above, you can choose to run all exercises on a local computer if you have pixi setup. You would then need to download data. We have tried to make the exercise data sets small such that you can run them locally.</p>
</section><section id="suggested-readings" class="level2"><h2 class="anchored" data-anchor-id="suggested-readings">Suggested readings</h2>
<p>Topics have been organized in modular units that consist of introductory slides and exercises. Some of the topics rely on concepts in population genetics and we recommend you at least skim these theoretical slides to be fully prepared for lectures.</p>
<p><a href="https://nbisweden.github.io/workshop-pgip-H25/slides/pgip">Population genomics in practice</a> provides a quick overview of population genomics. Although not required, it is recommended to briefly go through the following papers that will be discussed in the lecture:</p>
<ul>
<li>
<span class="citation" data-cites="fuller_PopulationGeneticsCoral_2020">Fuller et al. (<a href="#ref-fuller_PopulationGeneticsCoral_2020" role="doc-biblioref">2020</a>)</span> (most important)</li>
<li><span class="citation" data-cites="johri_RecommendationsImprovingStatistical_2022">Johri et al. (<a href="#ref-johri_RecommendationsImprovingStatistical_2022" role="doc-biblioref">2022</a>)</span></li>
</ul>
<p>See <a href="https://nbisweden.github.io/workshop-pgip-H25/slides/index.html#a-note-on-usage">notes on usage</a> for instructions how to view slides.</p>
</section><section id="resources" class="level2"><h2 class="anchored" data-anchor-id="resources">Resources</h2>
<section id="literature" class="level3"><h3 class="anchored" data-anchor-id="literature">Literature</h3>
<p>Lecture notes have been prepared based on the literature listed below.</p>
<section id="online" class="level4"><h4 class="anchored" data-anchor-id="online">Online</h4>
<dl>
<dt>Graham Coop’s notes on population genetics</dt>
<dd>
Comprehensive introduction to population genetics. Contains many biological examples and code snippets. <span class="citation" data-cites="grahamcoop_NotesPopulationGenetics_2020">(<a href="#ref-grahamcoop_NotesPopulationGenetics_2020" role="doc-biblioref">Graham Coop, 2020</a>)</span>.
</dd>
<dt>Jonathan Pritchard’s introduction to human population genetics.</dt>
<dd>
Although focussing more on human data, contains excellent explanations of statistical concepts and lots of instructive drawings. <span class="citation" data-cites="pritchard_OwnerGuideHuman_">(<a href="#ref-pritchard_OwnerGuideHuman_" role="doc-biblioref">Pritchard, n.d.</a>)</span>
</dd>
<dt>Joachim Hermisson’s notes on mathematical population genetics</dt>
<dd>
Introduction to mathematical population genetics <span class="citation" data-cites="hermisson_MathematicalPopulationGenetics_2017 hermisson_MathematicalPopulationGenetics_2018">(<a href="#ref-hermisson_MathematicalPopulationGenetics_2017" role="doc-biblioref">Hermisson, 2017</a>, <a href="#ref-hermisson_MathematicalPopulationGenetics_2018" role="doc-biblioref">2018</a>)</span>.
</dd>
</dl></section><section id="books" class="level4"><h4 class="anchored" data-anchor-id="books">Books</h4>
<dl>
<dt>Population Genetics: A Concise Guide</dt>
<dd>
John Gillespie’s short but excellent introduction to population genetics <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#ref-gillespie_PopulationGeneticsConcise_2004" role="doc-biblioref">Gillespie, 2004</a>)</span>.
</dd>
<dt>Molecular Population Genetics</dt>
<dd>
A more recent introduction to population genetics with more focus on the analyses of sequencing data <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#ref-hahn_MolecularPopulationGenetics_2019" role="doc-biblioref">Hahn, 2019</a>)</span>.
</dd>
<dt>Molecular Evolution and Phylogenetics</dt>
<dd>
Overview of molecular evolution and population genetics, and also phylogenetics <span class="citation" data-cites="nei_MolecularEvolutionPhylogenetics_2000">(<a href="#ref-nei_MolecularEvolutionPhylogenetics_2000" role="doc-biblioref">Nei &amp; Kumar, 2000</a>)</span>.
</dd>
<dt>Mathematical Population Genetics I</dt>
<dd>
A great reference when it comes to the mathematical treatment of population genetics <span class="citation" data-cites="ewens_MathematicalPopulationGenetics_2004">(<a href="#ref-ewens_MathematicalPopulationGenetics_2004" role="doc-biblioref">Ewens, 2004</a>)</span>.
</dd>
<dt>Principles of Population Genetics</dt>
<dd>
A comprehensive textbook covering most topics of population genetics <span class="citation" data-cites="hartl1997principles">(<a href="#ref-hartl1997principles" role="doc-biblioref">Hartl &amp; Clark, 1997</a>)</span>.
</dd>
<dt>Elements of Evolutionary Genetics</dt>
<dd>
Introduction to evolutionary genetics <span class="citation" data-cites="charlesworth_ElementsEvolutionaryGenetics_2010">(<a href="#ref-charlesworth_ElementsEvolutionaryGenetics_2010" role="doc-biblioref">Charlesworth &amp; Charlesworth, 2010</a>)</span>.
</dd>
<dt>Evolution</dt>
<dd>
Great comprehensive textbook covering evolution <span class="citation" data-cites="barton2007evolution">(<a href="#ref-barton2007evolution" role="doc-biblioref">Barton et al., 2007</a>)</span>.
</dd>
<dt>Coalescent Theory: An Introduction</dt>
<dd>
A great introduction to coalescent theory <span class="citation" data-cites="wakeley_CoalescentTheoryIntroduction_2008">(<a href="#ref-wakeley_CoalescentTheoryIntroduction_2008" role="doc-biblioref">Wakeley, 2008</a>)</span>.
</dd>
<dt>Gene Genealogies, Variation and Evolution: A Primer in Coalescent Theory</dt>
<dd>
Alternative introduction to coalescent with more focus on the Wright-Fisher model <span class="citation" data-cites="hein2005gene">(<a href="#ref-hein2005gene" role="doc-biblioref">Hein et al., 2005</a>)</span>.
</dd>
</dl>



</section></section></section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-barton2007evolution" class="csl-entry" role="listitem">
Barton, N. H., Briggs, D. E. G., Eisen, J. A., Goldstein, D. B., &amp; Patel, N. H. (2007). <em>Evolution</em>. <span>Cold Spring Harbor Laboratory Press</span>.
</div>
<div id="ref-charlesworth_ElementsEvolutionaryGenetics_2010" class="csl-entry" role="listitem">
Charlesworth, B., &amp; Charlesworth, D. (2010). <em>Elements of <span>Evolutionary Genetics</span></em>. <span>Roberts and Company Publishers</span>.
</div>
<div id="ref-ewens_MathematicalPopulationGenetics_2004" class="csl-entry" role="listitem">
Ewens, W. J. (2004). <em>Mathematical <span>Population Genetics</span></em> (S. S. Antman, J. E. Marsden, L. Sirovich, &amp; S. Wiggins, Eds.; Vol. 27). <span>Springer</span>. <a href="https://doi.org/10.1007/978-0-387-21822-9">https://doi.org/10.1007/978-0-387-21822-9</a>
</div>
<div id="ref-fuller_PopulationGeneticsCoral_2020" class="csl-entry" role="listitem">
Fuller, Z. L., Mocellin, V. J. L., Morris, L. A., Cantin, N., Shepherd, J., Sarre, L., Peng, J., Liao, Y., Pickrell, J., Andolfatto, P., Matz, M., Bay, L. K., &amp; Przeworski, M. (2020). Population genetics of the coral <span>Acropora</span> millepora: <span>Toward</span> genomic prediction of bleaching. <em>Science</em>, <em>369</em>(6501), eaba4674. <a href="https://doi.org/10.1126/science.aba4674">https://doi.org/10.1126/science.aba4674</a>
</div>
<div id="ref-gillespie_PopulationGeneticsConcise_2004" class="csl-entry" role="listitem">
Gillespie, J. H. (2004). <em>Population <span>Genetics</span>: <span>A Concise Guide</span></em> (2nd edition). <span>Johns Hopkins University Press</span>.
</div>
<div id="ref-grahamcoop_NotesPopulationGenetics_2020" class="csl-entry" role="listitem">
Graham Coop. (2020). <em>Notes on <span>Population Genetics</span></em>. <a href="https://github.com/cooplab/popgen-notes">https://github.com/cooplab/popgen-notes</a>
</div>
<div id="ref-hahn_MolecularPopulationGenetics_2019" class="csl-entry" role="listitem">
Hahn, M. (2019). <em>Molecular <span>Population Genetics</span></em> (First). <span>Oxford University Press</span>.
</div>
<div id="ref-hartl1997principles" class="csl-entry" role="listitem">
Hartl, D. L., &amp; Clark, A. G. (1997). <em>Principles of population genetics</em>. <span>Sinauer Associates</span>.
</div>
<div id="ref-hein2005gene" class="csl-entry" role="listitem">
Hein, J., Schierup, M. H., &amp; Wiuf, C. (2005). <em>Gene genealogies, variation and evolution: <span>A</span> primer in coalescent theory</em>. <span>Oxford University Press</span>. <a href="https://books.google.se/books?id=CCmLNAEACAAJ">https://books.google.se/books?id=CCmLNAEACAAJ</a>
</div>
<div id="ref-hermisson_MathematicalPopulationGenetics_2017" class="csl-entry" role="listitem">
Hermisson, J. (2017). <em>Mathematical population genetics</em>. <a href="https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017">https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017</a>
</div>
<div id="ref-hermisson_MathematicalPopulationGenetics_2018" class="csl-entry" role="listitem">
Hermisson, J. (2018). <em>Mathematical population genetics <span>II</span></em>. <a href="https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2018">https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2018</a>
</div>
<div id="ref-johri_RecommendationsImprovingStatistical_2022" class="csl-entry" role="listitem">
Johri, P., Aquadro, C. F., Beaumont, M., Charlesworth, B., Excoffier, L., Eyre-Walker, A., Keightley, P. D., Lynch, M., McVean, G., Payseur, B. A., Pfeifer, S. P., Stephan, W., &amp; Jensen, J. D. (2022). Recommendations for improving statistical inference in population genomics. <em>PLOS Biology</em>, <em>20</em>(5), e3001669. <a href="https://doi.org/10.1371/journal.pbio.3001669">https://doi.org/10.1371/journal.pbio.3001669</a>
</div>
<div id="ref-nei_MolecularEvolutionPhylogenetics_2000" class="csl-entry" role="listitem">
Nei, M., &amp; Kumar, S. (2000). <em>Molecular <span>Evolution</span> and <span>Phylogenetics</span></em>. <span>Oxford University Press</span>.
</div>
<div id="ref-pritchard_OwnerGuideHuman_" class="csl-entry" role="listitem">
Pritchard, J. K. (n.d.). <em>An <span>Owner</span>’s <span>Guide</span> to the <span>Human Genome</span></em>. Retrieved August 18, 2024, from <a href="https://web.stanford.edu/group/pritchardlab/HGbook.html">https://web.stanford.edu/group/pritchardlab/HGbook.html</a>
</div>
<div id="ref-wakeley_CoalescentTheoryIntroduction_2008" class="csl-entry" role="listitem">
Wakeley, J. (2008). <em>Coalescent <span>Theory</span>: <span>An Introduction</span></em> (1st Edition edition). <span>Roberts and Company Publishers</span>.
</div>
</div></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/NBISweden\.github\.io\/workshop-pgip\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Published with <a href="https://quarto.org/">Quarto</a> v<!--?quarto.version ?--></p>
</div>
  </div>
</footer>


</body></html>